### PR TITLE
Improve problem view performance

### PR DESCRIPTION
**Pull request by hyy215**
#### What it does

Fixed [#12407](https://github.com/eclipse-theia/theia/issues/12407)
1. Delay inserting nodes to marker tree. Collect and sort the nodes, and insert to the root.
If insert N nodes one by one, it maybe compare 1 + 2 + ... +  (N - 2) + (N - 1) + N times.
2. Creates a debounced function that delays updating decoration.
3. Setting markers for a model only if the monaco model exists.

#### How to test

Check that the bugs descriped in [#12407](https://github.com/eclipse-theia/theia/issues/12407). 
- Operating System: win10(cpu i7-8700K, memory 16G)

|number of files|before|now|
|---|---|---|
|500|4639ms|85ms|
|1000|17180ms|122ms|
|2000|69509ms|217ms|
|10000|too long|1160ms|

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/markers/src/browser/problem/problem-decorations-provider.ts [hunk 1]
```diff
@@ -35,13 +36,15 @@ export class ProblemDecorationsProvider implements DecorationsProvider {
...
-            this.currentUris = newUris;
-        });
+        this.problemManager.onDidChangeMarkers(() => this.fireDidDecorationsChanged());
     }
 
+    private fireDidDecorationsChanged = debounce(() => {
+        const newUris = Array.from(this.problemManager.getUris(), stringified => new URI(stringified));
+        this.onDidChangeEmitter.fire(newUris.concat(this.currentUris));
+        this.currentUris = newUris;
+    }, 10);
```
Comment (msujew) [inline packages/markers/src/browser/problem/problem-decorations-provider.ts:46]: Suggestion: A delay of 10ms is fairly short, and might not yield the expected improvements on a poor network connection. We should increase this delay to 50ms, which is in line with most other delays that we have in the frontend.

File: packages/markers/src/browser/problem/problem-tree-model.ts [hunk 2]
```diff
@@ -77,12 +79,21 @@ export class ProblemTree extends MarkerTree<Diagnostic> {
...
+        ProblemCompositeTreeNode.addChildren(this.root as MarkerRootNode, this._markers);
+
+        for (const { node, markers } of this._markers.values()) {
+            const children = this.getMarkerNodes(node, markers);
+            node.numberOfMarkers = markers.length;
+            this.setChildren(node, children);
+        }
+
+        this._markers.clear();
+    }, 10);
```
Comment (msujew) [inline packages/markers/src/browser/problem/problem-tree-model.ts:96]: Suggestion: We should use 50ms, see comment above.

File: packages/markers/src/browser/problem/problem-composite-tree-node.ts [hunk 3]
```diff
@@ -33,33 +33,46 @@ export namespace ProblemCompositeTreeNode {
         parent.severity = maxSeverity;
     };
 
-    export function addChild(parent: CompositeTreeNode, child: MarkerInfoNode, markers: Marker<Diagnostic>[]): CompositeTreeNode {
-        ProblemCompositeTreeNode.setSeverity(child, markers);
+    export function addChildren(parent: CompositeTreeNode, insertChildren: Map<string, { node: MarkerInfoNode, markers: Marker<Diagnostic>[] }>): void {
```
Comment (msujew) [inline packages/markers/src/browser/problem/problem-composite-tree-node.ts:36]: Question: What is the significance of storing the children in a map, if we only access the values of the map in this function? IMO we should just use a plain array in that case.

Comment (hyy215) [reply packages/markers/src/browser/problem/problem-composite-tree-node.ts:36]: Array is slower than Map in some cases. But in this case, they take about the same amount of time. I will use array.

Comment (hyy215) [reply packages/markers/src/browser/problem/problem-composite-tree-node.ts:36]: > Question: What is the significance of storing the children in a map, if we only access the values of the map in this function? IMO we should just use a plain array in that case.

Array is slower than Map in some cases. But in this case, they take about the same amount of time. I will use array.

**Review by msujew — APPROVED** [review]
Thank you for the contribution, looks good to me 👍

**Pull request merged**